### PR TITLE
Implement KMS handlers

### PR DESCRIPTION
**Pull request by reivaj05**
## Description
Implements KMS handlers

## Motivation and Context
We need to expose KMS logic in minio so it can be consumed in console

## How to test this PR?
Connect minio with [KES](https://github.com/minio/kes/wiki/MinIO-Object-Storage)

`curl --location --request POST '{MINIO_HOST}/api/v1/kms/key' \
--header 'Cookie: token={TOKEN}' \
--header 'Content-Type: application/json' \
--data-raw '{
    "key": "minio-test-key"
}'`

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
| mint-large-bucket.sh | :x: [more...](#09cd1fab-224c-46d7-96a9-299a867fb2f8) |
| mint-fs.sh | :x: [more...](#944de74a-864b-4f15-bb30-8ad2f4cc61da) |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#be0ecda2-f5dc-4514-a385-5d4ffac2645d) |
<a name="944de74a-864b-4f15-bb30-8ad2f4cc61da"></a>
### 15737-62e5657/mint-fs.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s1:31934
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 7 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 4 seconds
(5/14) Running awscli tests ... done in 56 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 24 seconds
(8/14) Running minio-go tests ... done in 53 seconds
(9/14) Running minio-java tests ... FAILED in 58 seconds
{
  "name": "minio-java",
  "function": "listObjects()",
  "args": "[bucket, prefix, recursive, 1050 objects]",
  "duration": 54861,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = BucketNotEmpty, message = The bucket you tried to delete is not empty, bucketName = minio-java-test-veeuv4, objectName = null, resource = /minio-java-test-veeuv4, requestId = 17173F0050C27ADC, hostId = 9db39bec-ac63-4309-bb8a-d42f39c441a3)\nrequest={method=DELETE, url=http://minio-k8s1:31934/minio-java-test-veeuv4, headers=Host: minio-k8s1:31934\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.4.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20220922T173057Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20220922/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=409, headers=Accept-Ranges: bytes\nContent-Length: 326\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nStrict-Transport-Security: max-age=31536000; includeSubDomains\nVary: Origin\nVary: Accept-Encoding\nX-Amz-Request-Id: 17173F0050C27ADC\nX-Content-Type-Options: nosniff\nX-Xss-Protection: 1; mode=block\nDate: Thu, 22 Sep 2022 17:30:57 GMT\n}\n >>> [io.minio.S3Base$1.onResponse(S3Base.java:690), okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519), java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128), java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628), java.base/java.lang.Thread.run(Thread.java:829)]"
}
(9/14) Running minio-js tests ... done in 58 seconds
(10/14) Running minio-py tests ... done in 46 seconds
(11/14) Running s3cmd tests ... done in 22 seconds
(12/14) Running s3select tests ... done in 1 seconds
(13/14) Running versioning tests ... done in 1 minutes and 1 seconds

Executed 13 out of 14 tests successfully.
```
<a name="be0ecda2-f5dc-4514-a385-5d4ffac2645d"></a>
### 15737-62e5657/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s1:32286
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 1 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 44 seconds
(4/14) Running aws-sdk-ruby tests ... done in 5 seconds
(5/14) Running awscli tests ... done in 57 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 28 seconds
(8/14) Running minio-go tests ... FAILED in 1 minutes and 15 seconds
{
  "args": {
    "bucketName": "minio-go-test-o0disx6w2s05szz9",
    "expires": 3600000000000,
    "objectName": "minio-go-test-o0disx6w2s05szz9unique",
    "reqParams": {
      "response-content-disposition": [
        "attachment; filename=\"test.txt\""
      ]
    }
  },
  "duration": 35771,
  "error": "Get \"http://minio-k8s1:32286/minio-go-test-o0disx6w2s05szz9/minio-go-test-o0disx6w2s05szz9unique?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=minio%2F20220922%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20220922T173007Z&X-Amz-Expires=3600&X-Amz-SignedHeaders=host&response-content-disposition=attachment%3B%20filename%3D%22test.txt%22&X-Amz-Signature=65041a1f263a5986129bd0010a330cf724269add9fc40029b8eb200b389dcc94\": context deadline exceeded (Client.Timeout exceeded while awaiting headers)",
  "function": "PresignedGetObject(bucketName, objectName, expires, reqParams)",
  "message": "PresignedGetObject response incorrect",
  "name": "minio-go: testFunctional",
  "status": "FAIL"
}
(8/14) Running minio-java tests ... done in 48 seconds
(9/14) Running minio-js tests ... done in 1 minutes and 2 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 26 seconds
(11/14) Running s3cmd tests ... done in 24 seconds
(12/14) Running s3select tests ... done in 4 seconds
(13/14) Running versioning tests ... done in 3 minutes and 4 seconds

Executed 13 out of 14 tests successfully.
```
<a name="09cd1fab-224c-46d7-96a9-299a867fb2f8"></a>
### 15737-62e5657/mint-large-bucket.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-k8s2:32222
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp :/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 9 seconds
(2/14) Running aws-sdk-java tests ... done in 1 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 6 seconds
(5/14) Running awscli tests ... done in 52 seconds
(6/14) Running healthcheck tests ... done in 1 seconds
(7/14) Running mc tests ... done in 27 seconds
(8/14) Running minio-go tests ... done in 1 minutes and 11 seconds
(9/14) Running minio-java tests ... FAILED in 2 minutes and 19 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[negative match etag]",
  "duration": 110119,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = NoSuchBucket, message = The specified bucket does not exist, bucketName = minio-java-test-1cqr0f2, objectName = null, resource = /minio-java-test-1cqr0f2, requestId = 17173F0D8920EBA2, hostId = 567158de-35c0-4580-ab1a-af91f658aa26)\nrequest={method=DELETE, url=http://minio-k8s2:32222/minio-java-test-1cqr0f2, headers=Host: minio-k8s2:32222\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.4.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20220922T173004Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20220922/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=404, headers=Accept-Ranges: bytes\nContent-Length: 318\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nServer: MinIO\nStrict-Transport-Security: max-age=31536000; includeSubDomains\nVary: Origin\nVary: Accept-Encoding\nX-Amz-Request-Id: 17173F0D8920EBA2\nX-Content-Type-Options: nosniff\nX-Xss-Protection: 1; mode=block\nDate: Thu, 22 Sep 2022 17:31:54 GMT\n}\n >>> [io.minio.S3Base$1.onResponse(S3Base.java:690), okhttp3.internal.connection.RealCall$AsyncCall.run(RealCall.kt:519), java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128), java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628), java.base/java.lang.Thread.run(Thread.java:829)]"
}
(9/14) Running minio-js tests ... done in 1 minutes and 3 seconds
(10/14) Running minio-py tests ... done in 1 minutes and 24 seconds
(11/14) Running s3cmd tests ... done in 25 seconds
(12/14) Running s3select tests ... done in 4 seconds
(13/14) Running versioning tests ... done in 3 minutes and 5 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/kms-router.go
Comment (klauspost) [inline cmd/kms-router.go]: Do we need this to be exported?
```suggestion
type kmsAPIHandlers struct{}
```

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/kms-router.go
Comment (klauspost) [inline cmd/kms-router.go]: Use lower case for function vars:

```suggestion
	kmsAPI := KMSAPIHandlers{}
```

etc... Increases readability. Upper case is for package level exported things.

File: cmd/kms-handlers.go
Comment (klauspost) [inline cmd/kms-handlers.go]: This 'smells' more like a GET style call, since it retrieves data and doesn't update it.

**Review by klauspost — COMMENTED** [review]
Some style comments.

For the listing operations, should we add limits/continue after functionality?

At some point we will probably need it, so not sure if we should add it already?

**Comment by harshavardhana** [comment]
> Some style comments.
> 
> For the listing operations, should we add limits/continue after functionality?
> 
> At some point we will probably need it, so not sure if we should add it already?

Pagination is perhaps not present in the Kes handler itself. We should look into that..

**Comment by reivaj05** [comment]
@klauspost Could you please help me to review it one more time?

I believe no pagination should be fine for now but i created a ticket to keep track of it and implement it later: https://github.com/miniohq/engineering/issues/980

**Comment by harshavardhana** [comment]
PTAL @aead

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

File: cmd/admin-router.go
Comment (harshavardhana) [inline cmd/admin-router.go]: you don't have to change the existing APIs you should be only concerned with kms-router.go @reivaj05

**Review by harshavardhana — CHANGES_REQUESTED** [review]

**Review by harshavardhana — APPROVED** [review]

**Comment by harshavardhana** [comment]
@aead can you PTAL this is waiting on your approval mainly!

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :heavy_check_mark: |
Deleting image on docker hub
Deleting image locally

**Review by klauspost — APPROVED** [review]
LGTM

**Review by aead — APPROVED** [review]

**Pull request merged**
